File: JavaScript/src/Internals/Functions/MathBase/Function_TRUNC.js
import { Function_2 } from '../Function_2.js';
import { Operand } from '../../../Operand.js';

class Function_TRUNC extends Function_2 {
    get Name() {
        return "Trunc";
    }

    constructor(z) {
        super(z);
    }

    evaluate(engine, tempParameter) {
        let args1 = this.getNumber_1(engine, tempParameter);
        if (args1.IsError) { return args1; }

        let digits = 0;
        if (this.b != null) {
            let args2 = this.getNumber_2(engine, tempParameter);
            if (args2.IsError) { return args2; }
            digits = args2.IntValue;
        }

        if (args1.NumberValue == 0) {
            return args1;
        }

        let num = args1.NumberValue;
        let factor = Math.pow(10, digits);

        let result;
        if (num > 0) {
            result = Math.floor(num * factor) / factor;
        } else {
            result = Math.ceil(num * factor) / factor;
        }
        return Operand.Create(result);
    }
}

export { Function_TRUNC };

File: java/toolgood.algorithm/src/main/java/toolgood/algorithm/internals/functions/mathBase/Function_TRUNC.java
package toolgood.algorithm.internals.functions.mathBase;

import java.math.BigDecimal;
import java.util.List;
import java.util.function.BiFunction;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.functions.NoneEngine;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_2;

final class Function_TRUNC extends Function_2 {
    public Function_TRUNC(FunctionBase[] funcs) {
        super(funcs);
        if (funcs.length < 1 || funcs.length > 2) {
            throw new IllegalArgumentException("Function '" + Name() + "' requires 1 to 2 parameters.");
        }
    }

    @Override
    public String Name() {
        return "Trunc";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        Operand args1 = GetNumber_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) {
            return args1;
        }
        int digits = 0;
        if (func2 != null) {
            Operand args2 = GetNumber_2(engine, tempParameter);
            if (args2.IsErrorOrNone()) {
                return args2;
            }
            digits = args2.IntValue();
        }
        if (args1.NumberValue().compareTo(BigDecimal.ZERO) == 0) {
            return args1;
        }
        BigDecimal num = args1.NumberValue();
        double factor = Math.pow(10, digits);
        double result;
        if (num.compareTo(BigDecimal.ZERO) > 0) {
            result = Math.floor(num.doubleValue() * factor) / factor;
        } else {
            result = Math.ceil(num.doubleValue() * factor) / factor;
        }
        return Operand.Create(BigDecimal.valueOf(result));
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.NUMBER;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        if (func2 != null) {
            func2.GetParameterTypes(noneEngine, result, OperandType.NUMBER);
        }
    }
}
